File: sdks/go/transport/factory.go
package transport

import (
	"fmt"
	"os"
	"strings"
	"time"
)

// BuildOptions controls how [Build] picks a transport.
type BuildOptions struct {
	// BaseURL — endpoint URL (`nexus://host:15475`, `http://host:15474`, …).
	// Defaults to `nexus://127.0.0.1:15475` when empty.
	BaseURL string
	// Transport is the explicit mode hint. URL scheme wins if set.
	Transport Mode
	// RpcPort overrides the default RPC port (15475).
	RpcPort uint16
	// Resp3Port overrides the default RESP3 port (15476).
	Resp3Port uint16
	// Timeout — HTTP request timeout (ignored by RPC).
	Timeout time.Duration
	// EnvTransport — injected test shim for NEXUS_SDK_TRANSPORT. Leave
	// empty to read from os.Environ.
	EnvTransport string
}

// Built is the resolved-transport tuple.
type Built struct {
	Transport Transport
	Endpoint  Endpoint
	Mode      Mode
}

// Build applies the precedence chain and returns a fresh transport.
//
// Precedence (highest wins):
//  1. URL scheme in BaseURL (`nexus://` → RPC, `http://` → HTTP, …)
//  2. NEXUS_SDK_TRANSPORT env var
//  3. BuildOptions.Transport field
//  4. Default: ModeNexusRpc
func Build(opts BuildOptions, creds Credentials) (Built, error) {
	var endpoint Endpoint
	if opts.BaseURL != "" {
		ep, err := ParseEndpoint(opts.BaseURL)
		if err != nil {
			return Built{}, err
		}
		endpoint = ep
	} else {
		endpoint = DefaultLocalEndpoint()
	}

	// 1. URL scheme wins.
	mode := schemeToMode(endpoint.Scheme)

	// 2. Env var overrides a bare URL (no scheme).
	explicitScheme := opts.BaseURL != "" && strings.Contains(opts.BaseURL, "://")
	envRaw := opts.EnvTransport
	if envRaw == "" {
		envRaw = os.Getenv("NEXUS_SDK_TRANSPORT")
	}
	if envMode, ok := ParseMode(envRaw); ok && !explicitScheme {
		mode = envMode
		endpoint = realignEndpoint(endpoint, mode, opts)
	}

	// 3. Config hint.
	if opts.Transport != "" && !explicitScheme && envRaw == "" {
		if _, ok := ParseMode(string(opts.Transport)); ok {
			mode = opts.Transport
			endpoint = realignEndpoint(endpoint, mode, opts)
		}
	}

	switch mode {
	case ModeNexusRpc:
		return Built{
			Transport: NewRpcTransport(endpoint, creds),
			Endpoint:  endpoint,
			Mode:      mode,
		}, nil
	case ModeHttp, ModeHttps:
		return Built{
			Transport: NewHttpTransport(endpoint, creds, opts.Timeout),
			Endpoint:  endpoint,
			Mode:      mode,
		}, nil
	case ModeResp3:
		return Built{}, fmt.Errorf(
			"resp3 transport is not yet shipped in the Go SDK — use 'nexus' (RPC) or 'http' for now",
		)
	}
	return Built{}, fmt.Errorf("unknown transport mode: %s", mode)
}

func schemeToMode(scheme string) Mode {
	switch scheme {
	case "nexus":
		return ModeNexusRpc
	case "resp3":
		return ModeResp3
	case "https":
		return ModeHttps
	default:
		return ModeHttp
	}
}

func realignEndpoint(ep Endpoint, mode Mode, opts BuildOptions) Endpoint {
	switch mode {
	case ModeNexusRpc:
		port := opts.RpcPort
		if port == 0 {
			port = RpcDefaultPort
		}
		return Endpoint{Scheme: "nexus", Host: ep.Host, Port: port}
	case ModeResp3:
		port := opts.Resp3Port
		if port == 0 {
			port = Resp3DefaultPort
		}
		return Endpoint{Scheme: "resp3", Host: ep.Host, Port: port}
	case ModeHttps:
		return Endpoint{Scheme: "https", Host: ep.Host, Port: HttpsDefaultPort}
	}
	return Endpoint{Scheme: "http", Host: ep.Host, Port: HttpDefaultPort}
}

File: sdks/python/nexus_sdk/transport/factory.py
"""Transport factory — applies the precedence chain.

URL scheme  >  NEXUS_SDK_TRANSPORT env var  >  config field  >  default (nexus)
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Optional

from nexus_sdk.transport.endpoint import (
    Endpoint,
    HTTP_DEFAULT_PORT,
    HTTPS_DEFAULT_PORT,
    RESP3_DEFAULT_PORT,
    RPC_DEFAULT_PORT,
    default_local_endpoint,
    parse_endpoint,
)
from nexus_sdk.transport.http_transport import HttpTransport
from nexus_sdk.transport.rpc import RpcTransport
from nexus_sdk.transport.types import (
    Transport,
    TransportCredentials,
    TransportMode,
)


@dataclass
class BuiltTransport:
    transport: Transport
    endpoint: Endpoint
    mode: TransportMode


def build_transport(
    base_url: Optional[str],
    credentials: TransportCredentials,
    transport_hint: Optional[TransportMode] = None,
    *,
    rpc_port: Optional[int] = None,
    resp3_port: Optional[int] = None,
    timeout_s: float = 30.0,
    env_transport: Optional[str] = None,
) -> BuiltTransport:
    """Resolve the effective transport given the precedence chain.

    ``env_transport`` is injected so the tests can exercise the env-var
    path without mutating ``os.environ``. Defaults to
    ``NEXUS_SDK_TRANSPORT``.
    """
    endpoint = parse_endpoint(base_url) if base_url else default_local_endpoint()

    # 1. URL scheme wins.
    mode = _scheme_to_mode(endpoint.scheme)

    # 2. Env var overrides a bare URL (no scheme).
    explicit_scheme = base_url is not None and "://" in base_url
    env_raw = env_transport if env_transport is not None else os.environ.get("NEXUS_SDK_TRANSPORT")
    env_mode = TransportMode.parse(env_raw) if env_raw else None
    if env_mode and not explicit_scheme:
        mode = env_mode
        endpoint = _realign_endpoint(endpoint, mode, rpc_port, resp3_port)

    # 3. Config field.
    if transport_hint and not explicit_scheme and not env_mode:
        mode = transport_hint
        endpoint = _realign_endpoint(endpoint, mode, rpc_port, resp3_port)

    if mode is TransportMode.NEXUS:
        return BuiltTransport(
            transport=RpcTransport(endpoint, credentials),
            endpoint=endpoint,
            mode=mode,
        )
    if mode in (TransportMode.HTTP, TransportMode.HTTPS):
        return BuiltTransport(
            transport=HttpTransport(endpoint, credentials, timeout_s=timeout_s),
            endpoint=endpoint,
            mode=mode,
        )
    if mode is TransportMode.RESP3:
        raise ValueError(
            "resp3 transport is not yet shipped in the Python SDK — use 'nexus' (RPC) "
            "or 'http' for now"
        )
    raise ValueError(f"unknown transport mode: {mode}")


def _scheme_to_mode(scheme: str) -> TransportMode:
    if scheme == "nexus":
        return TransportMode.NEXUS
    if scheme == "http":
        return TransportMode.HTTP
    if scheme == "https":
        return TransportMode.HTTPS
    if scheme == "resp3":
        return TransportMode.RESP3
    raise ValueError(f"unknown URL scheme: {scheme}")


def _realign_endpoint(
    ep: Endpoint,
    mode: TransportMode,
    rpc_port: Optional[int],
    resp3_port: Optional[int],
) -> Endpoint:
    """Retarget the port when the mode changes out from under the URL."""
    if mode is TransportMode.NEXUS:
        return Endpoint(scheme="nexus", host=ep.host, port=rpc_port or RPC_DEFAULT_PORT)
    if mode is TransportMode.RESP3:
        return Endpoint(scheme="resp3", host=ep.host, port=resp3_port or RESP3_DEFAULT_PORT)
    if mode is TransportMode.HTTPS:
        return Endpoint(scheme="https", host=ep.host, port=HTTPS_DEFAULT_PORT)
    return Endpoint(scheme="http", host=ep.host, port=HTTP_DEFAULT_PORT)
